javascript/ql/src/semmle/javascript/frameworks/Vue.qll

```diff
@@ -0,0 +1,380 @@
+
+  }
+
+  /**
+   * An extended Vue from `Vue.extend({...})`.
+   */
+  class ExtendedVue extends Instance, MkExtendedVue {
+    VueExtend extend;
+
+    ExtendedVue() { this = MkExtendedVue(extend) }
+
+    override string toString() { result = extend.toString() }
+
+    override predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    ) {
+      extend.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
+    }
+
+    override DataFlow::Node getOwnOption(string name) { result = extend.getOptionArgument(0, name) }
+
+  }
+
+  /**
+   * An instance of an extended Vue, for example `instance` of `var Ext = Vue.extend({...}); var instance = new Ext({...})`.
+   */
+  class ExtendedInstance extends Instance, MkExtendedInstance {
+    VueExtend extend;
+
+    DataFlow::NewNode sub;
+
+    ExtendedInstance() { this = MkExtendedInstance(extend, sub) }
+
+    override string toString() { result = sub.toString() }
+
+    override predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    ) {
+      sub.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
+    }
+
+    override DataFlow::Node getOwnOption(string name) { result = sub.getOptionArgument(0, name) }
+
+    override DataFlow::Node getOption(string name) {
+      result = Instance.super.getOption(name)
+      or
+      result = MkExtendedVue(extend).(ExtendedVue).getOption(name)
+    }
+
+  }
+
+  /**
+   * A Vue component from `Vue.component("my-component", { ... })`.
+   */
+  class Component extends Instance, MkComponent {
+    DataFlow::CallNode def;
+
+    Component() { this = MkComponent(def) }
+
+    override string toString() { result = def.toString() }
+
+    override predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    ) {
+      def.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
+    }
+
+    override DataFlow::Node getOwnOption(string name) { result = def.getOptionArgument(1, name) }
+
+  }
+
+  /**
+   * A single file Vue component in a `.vue` file.
+   */
+  class SingleFileComponent extends Instance, MkSingleFileComponent {
+    VueFile file;
+
+    SingleFileComponent() { this = MkSingleFileComponent(file) }
+
+    override predicate hasLocationInfo(
+      string filepath, int startline, int startcolumn, int endline, int endcolumn
+    ) {
+      filepath = file.getAbsolutePath() and
+      startline = 0 and
+      startcolumn = 0 and
+      endline = 0 and
+      endcolumn = 0
+    }
+
+    private Module getModule() {
+      exists(HTML::ScriptElement script | script.getFile() = file |
+        result.getTopLevel() = script.resolveSource()
+        or
+        not exists(script.getSourcePath()) and
+        // XXX is there no way explicit relation between a script tag and its inline script?
```

Comment:
@xiemaisi  this comment should obviously not be merged as is. Do we have a way of establishing the relation?

Reply:
You should be able to relate the two by their location.

Reply:
Sorry for asking again, but then we need to check for positional containment of one in the other, do we really want to do that?
The location of the `HTML::ScriptElement` is the range between the `<`/`>` characters, and the location of an associated `InlineScript` is the range inside `>`/`<`. For example: `HTML::ScriptElement` has location `test.html:1:1:1:21`, while `InlineScript` has location `test.html:1:9:1:12` for:

```
<script>code</script>
```

Reply:
I'm not sure what you are asking. I realise that it's not as simple as comparing two positions for equality, but it's not much more complicated either. The only reason I can see why we would _not_ want to do it this way is it if it turns out to be very difficult to scale.